File: Algo/Leetcode/050Powxn.cpp
class Solution {
public:
    double myPow(double x, int n) {
        if(n==0) return 1.0;

        if(isZero(x)) return 0.0;

        long n1 = n;
        if (n1<0){
            n1 *= -1;
            x = 1/x;
        }

        double res = 1.0;
        while(n1>0){
            if (n1 & 1) res *= x;
            x *= x;
            n1>>=1;
        }
        return res;
    }

    bool isZero(double x){
        return abs(x) <= 1e-15;
    }
};

File: Algo/Leetcode/050Powxn.py
class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if n == 0:
            return 1.0
        elif n == 1:
            return x
        elif n < 0:
            return self.myPow(1.0/x, -n)
        else:
            if n % 2 == 0:
                return self.myPow(x*x, n/2)
            else:
                return x * self.myPow(x*x, n/2)